platforms/ios/lib/WysiwygComposer/Sources/WysiwygComposer/Tools/StringDiffer.swift

```diff
@@ -0,0 +1,98 @@
+//
+// Copyright 2022 The Matrix.org Foundation C.I.C
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+import OSLog
+
+/// Describe an error occurring during string diffing.
+enum StringDifferError: LocalizedError, Equatable {
+    case tooComplicated
+    case insertionsDontMatchRemovals
+
+    var errorDescription: String? {
+        switch self {
+        case .tooComplicated:
+            return "Diff is too complicated to be handled as a simple replacement"
+        case .insertionsDontMatchRemovals:
+            return "Insertions don't match removals"
+        }
+    }
+}
+
+/// Describes the output replacement from string diffing.
+struct StringDifferReplacement: Equatable {
+    let range: NSRange
+    let text: String
+
+    init(range: NSRange, text: String) {
+        self.range = range
+        self.text = text
+    }
+
+    init(location: Int, length: Int, text: String) {
+        range = NSRange(location: location, length: length)
+        self.text = text
+    }
+}
+
+final class StringDiffer {
+    // MARK: - Private
+
+    private init() { }
+
+    // MARK: - Internal
+
+    static func replacement(from oldText: String, to newText: String) throws -> StringDifferReplacement? {
+        let difference = newText.withNBSP.difference(from: oldText.withNBSP)
+
+        guard !difference.isEmpty else {
+            return nil
+        }
+
+        let removedRanges = difference.removedRanges
+        let textInsertions = difference.textInsertions
+
+        guard removedRanges.count < 2, textInsertions.count < 2 else {
+            throw StringDifferError.tooComplicated
```

Comment:
Could a case like this where the diff would return multiple removed ranges and text insertions actually happen?

Reply:
Hopefully, it doesn't. We could be able to handle it, but I'd rather not waste time trying to match the removals to the correct insertions.